### PR TITLE
Fix `ir_utils._max_label` being updated incorrectly

### LINKED ISSUE #7155: Parfors max_label issue causing undefined behavior
A parfors error is revealed by the following code that uses the retargeting support:

```python
import numpy as np

from numba import njit
from numba.core.dispatcher import TargetConfig
from numba.core.target_extension import (
    dispatcher_registry,
    CPUDispatcher,
    CPU,
    target_registry,
)
from numba.core.retarget import BasicRetarget

# ------------ A "CustomCPU" target ------------


class CustomCPU(CPU):
    pass


class CustomCPUDispatcher(CPUDispatcher):
    pass


target_registry["CustomCPU"] = CustomCPU
dispatcher_registry[target_registry["CustomCPU"]] = CustomCPUDispatcher


# ------------ For switching target ------------


class MyCustomTarget(BasicRetarget):
    @property
    def output_target(self):
        return "CustomCPU"

    def compile_retarget(self, cpu_disp):
        kernel = njit(_target="CustomCPU", parallel=True)(cpu_disp.py_func)
        return kernel


retarget = MyCustomTarget()

# ------------ Functions being tested ------------
@njit
def f(a):
    return np.arange(a.size)


def main():
    a = np.ones(20)
    with TargetConfig.switch_target(retarget):
        r = f(a)


if __name__ == "__main__":
    main()
```

This code will lead to the following error as of 190237189fc59f99adeba2958b6c9cd0940dbf39.

```pytb
TypeError: can't unbox array from PyObject into native value.  The object maybe of a different type
```

With `NUMBA_DEBUG=1`, we can see that the optimized LLVM for `f` is:

```llvm
define i32 @"_ZN8__main__5f$241E5ArrayIdLi1E1C7mutable7alignedE"({ i8*, i8*, i64, i64, i64*, [1 x i64], [1 x i64] }* noalias nocapture readnone %retptr, { i8*, i32, i8* }** noalias nocapture readnone %excinfo, i8* nocapture readnone %arg.a.0, i8* nocapture readnone %arg.a.1, i64 %arg.a.2, i64 %arg.a.3, double* nocapture readnone %arg.a.4, i64 %arg.a.5.0, i64 %arg.a.6.0) local_unnamed_addr #0 {
entry:
  br label %B239.endif

B239.endif:                                       ; preds = %B239.endif, %entry
  br label %B239.endif
}
```

containing just a infinite loop.

After some debugging, I can observe that `ir_utils._max_label` is not being updated correctly. As a result, subsequent `ir_utils.next_label()` may return a preexisting label.

### PR DESCRIPTION
**Pull request by sklam**
Fixes #7155.

This patch changes `ir_utils._max_label` into an instance of `_MaxLabel` so it is easier to enforce the correct update logic. It also means that we no longer need to use `global _max_label` to update it.

**Additional notes**

As demonstrate by this patch,  #7155 is not caused by the retargeting support or any of the recent work to make target-specific overload. It is just unfortunate that the bug is revealed in a case that make use of the retargeting support. I'm also unsure if we really should keep the included testcase, which is not representative of the problem but just a coincident that it revealed the problem. It is likely overly difficult to create a minimal reproducer for the problem.

### CODE REVIEW

**Comment by sklam** [comment]
The tests are taking overly long because of high memory consumption. The problem is not unique to this PR branch. On mainline, running `python runtests.py numba/tests/test_parfors.py` quickly uses up >1GB on my machine.

**Comment by stuartarchibald** [comment]
> The tests are taking overly long because of high memory consumption. The problem is not unique to this PR branch. On mainline, running `python runtests.py numba/tests/test_parfors.py` quickly uses up >1GB on my machine.

Like this: https://github.com/numba/numba/issues/7028#issuecomment-844122335 ?

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch. Few minor queries else looks good. I plan on testing this manually next.

**Comment by stuartarchibald** [comment]
@sklam https://github.com/numba/numba/pull/7160 should sort out the issue with memory pressure from the parfors tests.

File: numba/core/ir_utils.py [hunk 1]
```diff
@@ -36,7 +36,19 @@ def mk_unique_var(prefix):
...
+
+    def next(self):
+        self._value += 1
+        return self._value
+
+    def update(self, newval):
+        self._value = max(newval, self._value)
+
+
+_the_max_label = _MaxLabel()
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/core/ir_utils.py:51]: sure

File: numba/core/ir_utils.py [hunk 2]
```diff
@@ -1662,10 +1672,9 @@ def compile_to_numba_ir(mk_func, glbls, typingctx=None, targetctx=None,
     remove_dels(f_ir.blocks)
 
     # relabel by adding an offset
-    global _max_label
-    f_ir.blocks = add_offset_to_labels(f_ir.blocks, _max_label + 1)
+    f_ir.blocks = add_offset_to_labels(f_ir.blocks, _the_max_label.next())
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/core/ir_utils.py:1675]: The counter not updated part had me worried. That's why i changed it to advance. Worse case is that there's a hole the label space, but I want to be certain that there is no reuse of label.

File: numba/core/inline_closurecall.py [hunk 3]
```diff
@@ -579,14 +579,14 @@ def inline_closure_call(func_ir, glbls, block, i, callee, typingctx=None,
...
-    max_label = max(ir_utils._max_label, max(func_ir.blocks.keys()))
+    max_label = max(ir_utils._the_max_label.next(), max(func_ir.blocks.keys()))
     callee_blocks = add_offset_to_labels(callee_blocks, max_label + 1)
     callee_blocks = simplify_CFG(callee_blocks)
     callee_ir.blocks = callee_blocks
     min_label = min(callee_blocks.keys())
     max_label = max(callee_blocks.keys())
     #    reset globals in ir_utils before we use it
-    ir_utils._max_label = max_label
+    ir_utils._the_max_label.update(max_label)
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/core/inline_closurecall.py:589]: One of the bugs. I stopped looking further for where exactly the problems are and just batch fixes all the `_max_label`

File: numba/tests/parfors_max_label_error.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+class CustomCPU(CPU):
+    pass
+
+
+class CustomCPUDispatcher(CPUDispatcher):
+    pass
+
+
+target_registry["CustomCPU"] = CustomCPU
+dispatcher_registry[target_registry["CustomCPU"]] = CustomCPUDispatcher
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/tests/parfors_max_label_error.py:34]: Right. The test is also highly dependent of the internal state of `ir_utils` that it must be executing in a fresh process to replicate.

File: numba/core/ir_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/core/ir_utils.py:1675]: Right, I agree, it seems like it ought to update else there is potential for reuse, and like you said, worse case is that the label space has gaps which is relatively harmless.

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes.

**Comment by sklam** [comment]
CI passed. Merging

**Pull request merged**